crates/pep508-rs/src/marker/tree.rs

```diff
@@ -1010,258 +934,451 @@ impl MarkerTree {
-                    negated.and(tree.negate());
-                }
-                negated
+        };
+
+        match string_marker.key() {
+            MarkerValueString::OsNameDeprecated => {
+                reporter.report(
+                    MarkerWarningKind::DeprecatedMarkerName,
+                    "os.name is deprecated in favor of os_name".to_string(),
+                );
+            }
+            MarkerValueString::PlatformMachineDeprecated => {
+                reporter.report(
+                    MarkerWarningKind::DeprecatedMarkerName,
+                    "platform.machine is deprecated in favor of platform_machine".to_string(),
+                );
+            }
+            MarkerValueString::PlatformPythonImplementationDeprecated => {
+                reporter.report(
+                    MarkerWarningKind::DeprecatedMarkerName,
+                    "platform.python_implementation is deprecated in favor of
+                        platform_python_implementation"
+                        .to_string(),
+                );
+            }
+            MarkerValueString::PythonImplementationDeprecated => {
+                reporter.report(
+                    MarkerWarningKind::DeprecatedMarkerName,
+                    "python_implementation is deprecated in favor of
+                        platform_python_implementation"
+                        .to_string(),
+                );
+            }
+            MarkerValueString::PlatformVersionDeprecated => {
+                reporter.report(
+                    MarkerWarningKind::DeprecatedMarkerName,
+                    "platform.version is deprecated in favor of platform_version".to_string(),
+                );
+            }
+            MarkerValueString::SysPlatformDeprecated => {
+                reporter.report(
+                    MarkerWarningKind::DeprecatedMarkerName,
+                    "sys.platform  is deprecated in favor of sys_platform".to_string(),
+                );
             }
+            _ => {}
+        }
+
+        for (_, tree) in string_marker.children() {
+            tree.report_deprecated_options(reporter);
         }
     }
 
-    /// Combine this marker tree with the one given via a conjunction.
+    /// Find a top level `extra == "..."` expression.
     ///
-    /// This does some shallow flattening. That is, if `self` is a conjunction
-    /// already, then `tree` is added to it instead of creating a new
-    /// conjunction.
-    pub fn and(&mut self, tree: MarkerTree) {
-        if self == &tree {
-            return;
-        }
-        match *self {
-            MarkerTree::Expression(_) | MarkerTree::Or(_) => {
-                let this = std::mem::replace(self, MarkerTree::And(vec![]));
-                *self = MarkerTree::And(vec![this]);
+    /// ASSUMPTION: There is one `extra = "..."`, and it's either the only marker or part of the
+    /// main conjunction.
+    pub fn top_level_extra(&self) -> Option<MarkerExpression> {
```

Comment:
I kept extras at the lowest level because it makes `simplify_extras` cheaper. Having extras at the top would mean `simplify_extras` has to create a new node for every single node in the tree (because nodes are immutable).

Also, extras are not represented as a single variable, each extra expression creates a distinct variable.